lib/cli/ui/prompt/interactive_options.rb

```diff
@@ -68,61 +79,96 @@ def call
 
         private
 
-        def reset_position
+        def calculate_option_line_lengths
+          @terminal_width_at_calculation_time = CLI::UI::Terminal.width
+          # options will be an array of questions but each option can be multi-line
+          # so to get the # of lines, you need to join then split
+
+          # since lines may be longer than the terminal is wide, we need to
+          # determine how many extra lines would be taken up by them
+          max_width = (@terminal_width_at_calculation_time -
+                       @options.count.to_s.size - # Width of the displayed number
+                       5 -                        # Extra characters added during rendering
+                       (@multiple ? 1 : 0)        # Space for the checkbox, if rendered
+                      ).to_f
+
+          @option_lengths = @options.map do |text|
+            width = 1 if text.empty?
+            width ||= text
+                        .split("\n")
+                        .reject(&:empty?)
+                        .map { |l| (l.length / max_width).ceil }
+                        .reduce(&:+)
+
+            width
+          end
+        end
+
+        def reset_position(number_of_lines=num_lines)
           # This will put us back at the beginning of the options
           # When we redraw the options, they will be overwritten
           CLI::UI.raw do
-            num_lines.times { print(ANSI.previous_line) }
+            number_of_lines.times { print(ANSI.previous_line) }
           end
         end
 
-        def clear_output
+        def clear_output(number_of_lines=num_lines)
           CLI::UI.raw do
             # Write over all lines with whitespace
-            num_lines.times { puts(' ' * CLI::UI::Terminal.width) }
+            number_of_lines.times { puts(' ' * CLI::UI::Terminal.width) }
           end
-          reset_position
-        end
+          reset_position number_of_lines
 
-        def num_lines
-          options = presented_options.map(&:first)
+          # Update if metadata is being displayed
+          # This must be done _after_ the output is cleared or it won't draw over
+          # the entire output
+          @displaying_metadata = display_metadata?
+        end
 
-          # empty_option_count is needed since empty option titles are omitted
-          # from the line count when reject(&:empty?) is called
-          empty_option_count = options.count(&:empty?)
+        # Don't use this in place of +@displaying_metadata+, this updates too
+        # quickly to be useful when drawing to the screen.
+        def display_metadata?
+          filtering? or selecting? or has_filter?
+        end
 
-          # options will be an array of questions but each option can be multi-line
-          # so to get the # of lines, you need to join then split
+        def num_lines
+          calculate_option_line_lengths if terminal_width_changed?
```

Comment:
Oh, cool. It never even occurred to me that it would be possible to mostly solve this problem.